Always pass explicit vibrate setting for notifications

beroset opened the pull request:
The current code always sets the vibrate level of notifications to
"Strong" no matter the current setting.  This change causes the client
to respect the currently set vibration settings instead.

beroset committed:
Remove SetVibration and add vibration to SendNotify

As the complementary change to the libasteroid library, this removes the
SetVibration call from the interface and adds an explicit vibration
parameter to SendNotify.  This has the effect that now all SendNotify
requests must explicitly set the vibration setting which makes more
sense than storing and not using a pre-set vibration value and
streamlines the interface.

FlorentRevest commented:
This is an improvement because it works but I think the current vibration API is totally wrong and now is a good time to fix it.

I think the vibration pattern should be naturally provided as an additional argument to `void DBusWatch::SendNotify(unsigned int id, QString appName, QString icon, QString body, QString summary)`.

The vibration pattern is a property of a notification not of the watch so the current `void DBusWatch::SetVibration(QString v)` makes no sense and should be deleted. Furthermore it currently just calls a libasteroid function which makes no sense, it only remembers `m_vibrate` but this is never used in libasteroid https://github.com/AsteroidOS/libasteroid/blob/5f7e9f2fe321fbbb076c252ba72dba9aaa064c71/src/services/notificationservice.cpp#L70

I think libasteroid should not be in the business of remembering what vibration settings should be used for all notifications and the process that issues notifications should know what vibration pattern is best for each notification individually. (for example, in AsteroidOSSync, on Android, we have a hashmap that remembers the preferred vibration pattern for each app)

beroset commented:
I agree with you that the better fix is to change the API to add the vibration setting to `DBusWatch::SendNotify`.  I'll revise the patch to do just that, but of course this will break the interface for current users.  For that reason, I'd like to bump the version to 2.0.0.

FlorentRevest commented:
> For that reason, I'd like to bump the version to 2.0.0

Sure, why not. Numbers mean what we want them to mean so if you think a number bump is appropriate, that's fine by me. Just keep in mind there are not so many users of this at the moment so it's still a good time to "move fast and break things" :)

beroset commented:
This depends on the https://github.com/AsteroidOS/libasteroid/pull/14 being merged first.  Once that's complete, I'll submit the revised PR for this project.

beroset committed:
Update SendNotify call for Ubuntu Touch to new API

Because the updated API now requires all callers to explicitly set the
vibration level for each SendNotify call, and because the libasteroid
library no longer supports getVibration(), the Ubuntu Touch platform now
always sets the vibration to "normal".  If a global setting is ever
created on that platform, this code could be modified to use that
setting rather than the hardcoded value.

beroset committed:
Bump version to 2.0.0 due to API changes

Since this version changes the interface by removing SetVibration() and
requiring a vibration parameter in SendNotify(), it makes sense to
update the major version number to allow this change to be detected by
software that uses asteroidsyncservice.

beroset committed:
Add vibration parameter to sendNotify()

This is the complementary change to the asteroidsyncservice plugin that
matches the change in API to require a vibration parameter for
sendNotify requests.

FlorentRevest commented on asteroidsyncserviced/platforms/ubuntutouch/platform.cpp:
```diff
@@ -62,7 +62,7 @@ uint Platform::Notify(const QString &app_name, uint replaces_id, const QString &
             qDebug() << __func__ << "Have a phone call notification. Ignoring it..." << app_name << app_icon;
         } else if(m_notificationServiceReady) {
             unsigned int randId = rand() % 100000 + 1;
-            m_notificationService->insertNotification("", randId, encodeAppName(app_name), encodeIcon(app_name), summary, body, m_notificationService->getVibration());
```
Could you also adjust asteroidsyncserviced/platforms/sailfishos/platform.cpp ?

FlorentRevest commented on asteroidsyncserviced/platforms/ubuntutouch/platform.cpp:
```diff
@@ -62,7 +62,7 @@ uint Platform::Notify(const QString &app_name, uint replaces_id, const QString &
         } else if(m_notificationServiceReady) {
             unsigned int randId = rand() % 100000 + 1;
-            m_notificationService->insertNotification("", randId, encodeAppName(app_name), encodeIcon(app_name), summary, body, m_notificationService->getVibration());
+            m_notificationService->insertNotification("", randId, encodeAppName(app_name), encodeIcon(app_name), summary, body, m_notificationService->strToVibrate("normal"));
```
If you call strToVibrate here only to convert it back to a string right away (in libasteroid). How about just forwarding a string from the beginning instead ? We could remove all the enum/string conversion logic from libasteroid too

beroset replied:
There are only four valid values and no way to indicate that via D-Bus.  An alternative would be to use an int instead, which I think would be neater.  Which do you prefer?

beroset replied:
Yes, they're all using QString now

FlorentRevest commented:
Also general comment, I see you didn't clean up the git history. It's not great when the git history reflects the back and forths of the PR review, ideally the git log should be straight to the point and reflect what is done so that people going through the log of the repo in the future won't have to go through the back and forths themselves.

beroset commented:
Since you had already commented on the earlier commit, I had thought that removing it entirely would be confusing, but on further reflection, these comments are ephemeral and the log is permanent, so I'll be more vigilant about the commit logs and not worry about the PR review comments.

FlorentRevest commented:
GitHub is smart enough to keep comments, attached to old commits :) so this should not be a concern.

One reason why I think that passing a string here is the right decision is cross-version compatibility and maintenance burden. Imagine we add a new vibration pattern to asteroid tomorrow, if you put a filter for allowed values in libasteroid that means we need to change the OS and the sync app AND libasteroid. But if libasteroid forwards a string and we rely on the watch to decide what to do with unknown values, then we know what happens with a recent sync app and an old asteroid (if you send an unrecognized value, asteroid itself will default to strong) and we also have an easier time upgrading sync apps :)

I see this is still marked as draft, I'm assuming this is awaiting the SailfishOS change equivalent to the Ubuntu change ? Otherwise the current state of things looks good to me :)

beroset committed:
Advance libasteroid to update notification service

This uses the newest version of libasteroid that has a change to the
notification service that uses a QString for the vibrate parameter.

beroset committed:
Fix sailfish to use updated notification service

This makes the corresponding change to sailfish to track the libasteroid
change that now uses a QString for the vibrate parameter.

beroset committed:
Merge branch 'AsteroidOS:master' into use-current-vibrate-setting

FlorentRevest commented:
Apologies it took me so long to find some free time for reviews! Thanks for that work :D
